### PR TITLE
[timeseries] Add native support for missing values

### LINKED ISSUE #3886: [timeseries] Add first-class support for missing values
Currently, `TimeSeriesPredictor` deals with missing values in the data by first imputing them via forward-/backward-filling, and then training all models as if there are no missing values. This strategy may lead to poor accuracy on datasets with a large portion of missing values since training data will include regions with constant values arising from ffilling. 

A better alternative is to keep the missing values represented by `nan` in the data, and let models handle it. This requires following modifications to the code:
- [x] Update metric implementations to handle missing values in target (#3966)
- [ ] Update preprocessing logic in `TimeSeriesPredictor._check_and_prepare_data_frame`
- [ ] Make sure that all models can handle missing values. This means, all models can train normally and produce forecasts with no NaN values, even if training data contains NaNs.
  - [ ] GluonTS models (DeepAR, TFT, PatchTST, DLinear)
  - [ ] StatsForecast models (AutoETS, AutoARIMA, Theta, intermittent demand models)
  - [ ] Baseline models (Naive, SeasonalNaive, Average, SeasonalAverage, Zero)
  - [ ] MLForecast models (DirectTabular, RecursiveTabular)

### PR DESCRIPTION
**Pull request by shchur**
*Issue #, if available:* fixes #3886 

*Description of changes:*
- Instead of imputing all missing values in `target` column in `TimeSeriesPredictor._check_and_prepare_data_frame`, we let each model use its own logic for handling missing values.
  - GluonTS models (DeepAR, TFT, PatchTST, DLinear) + some local models (Average, SeasonalAverage, NPTS, Naive, SeasonalNaive) handle the missing values natively
  - Other local models (AutoETS, AutoCES, AutoARIMA, Theta, intermittent models) perform imputation first
  - MLForecast models use a mix of two strategies: missing values are imputed, but rows that originally contained NaN values are not used for training
- Model properties (e.g., whether it can handle missing values) are stored using the `_get_tags()` mechanism
- `TimeSeriesPredictor` now removes time series consisting of only NaN values from `train_data` during `fit()`
- Missing values in covariates are still always imputed inside `TimeSeriesFeatureGenerator`
- Add missing values to `DUMMY_TS_DATAFRAME` used in the tests to ensure that NaN support works in all scenarios

*To do:*
- [x] Add tests
- [x] Add missing values support to MLForecast models
- [x] Benchmark the new NaN handling strategy on datasets with missing values

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| huggingface-hub==0.21.4 | huggingface-hub==0.22.0 |
| lightning-utilities==0.11.0 | lightning-utilities==0.11.1 |
| filelock==3.13.1 | filelock==3.13.2 |
| huggingface-hub==0.21.4 | huggingface-hub==0.22.0 |
| lightning-utilities==0.11.0 | lightning-utilities==0.11.1 |
| filelock==3.13.1 | filelock==3.13.2 |

File: timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py
Comment (canerturkmen) [inline timeseries/src/autogluon/timeseries/models/autogluon_tabular/mlforecast.py:95]: Just to check, this should never happen when called through the predictor?

File: timeseries/tests/unittests/models/test_models.py
Comment (canerturkmen) [inline timeseries/tests/unittests/models/test_models.py:506]: Just to check: We never modify the user-provided dataframe inplace, right?

**Review by canerturkmen — COMMENTED** [review]
LGTM overall thanks a lot for this! Dropped a few minor comments and questions.

File: timeseries/src/autogluon/timeseries/utils/features.py
Comment (shchur) [reply timeseries/src/autogluon/timeseries/utils/features.py:202]: The first line 
```python
ts_df[covariates_names] = ts_df[covariates_names].fill_missing_values()
```
does ffill + bfill for all covariates (per item). If there are still missing values left after that, it means that for some time series that covariate consists completely of NaNs. In this case, we use the median of the covariate in the dataset for imputation.

Thinking about it, we probably need to store the median covariate values of the training data during fit and use those for imputation. This would make more sense in case user tries to predict items one by one (and for some items the covariate is all-nan)

File: timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/abstract/abstract_timeseries_model.py:241]: Got it. Thanks!

File: timeseries/src/autogluon/timeseries/models/local/npts.py
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/models/local/npts.py:93]: To better articulate my question: if the target start position is not nan but the sampled trajectory does contain a NaN, would the forecast have NaNs?

File: timeseries/src/autogluon/timeseries/utils/features.py
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/utils/features.py:20]: It can still complicate things due to dependency version conflicts or just one more thing to maintain. I would rather not introduce a dependency for this sort of functionality, but your call.

File: timeseries/src/autogluon/timeseries/utils/features.py
Comment (canerturkmen) [reply timeseries/src/autogluon/timeseries/utils/features.py:202]: That's quite complicated logic. If we can get away with filling zeros for this edge case for example, I'm OK with that (with a loud warning as well) just to save complexity.

File: timeseries/src/autogluon/timeseries/utils/features.py
Comment (shchur) [reply timeseries/src/autogluon/timeseries/utils/features.py:20]: Fair enough, I guess this can be handled by a static type checker without introducing a dependency

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |
| flatbuffers==24.3.7 | flatbuffers==24.3.25 |
| flatbuffers==24.3.7 | flatbuffers==24.3.25 |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

File: timeseries/tests/unittests/test_features.py
Comment (canerturkmen) [inline timeseries/tests/unittests/test_features.py:158]: This looks great and I'll probably use it for feature importance.

Should we also test the filling logic, at least for the 'median' and 'mode' scenarios?

**Review by canerturkmen — APPROVED** [review]
LGTM! Only one comment.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Review by shchur — COMMENTED** [review]
Thanks @canerturkmen, I've added the tests for mode/median imputation + added missing values to `DUMMY_TS_DATAFRAME` so that we consider more settings with NaNs

**Comment by shchur** [comment]
@canerturkmen just finished some benchmarking on 12 datasets with missing values / 3 folds each: This PR branch has 62.5% win rate vs. current `master` branch, so I guess we are good to merge once the problems with tests are resolved.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

File: timeseries/tests/unittests/models/test_local.py
Comment (shchur) [inline timeseries/tests/unittests/models/test_local.py:89]: We already have the same test here https://github.com/autogluon/autogluon/blob/master/timeseries/tests/unittests/models/test_models.py#L269.

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Comment by yinweisu** [comment]
| Previous CI Run | Current CI Run |
| :---: | :---: |

**Review by canerturkmen — APPROVED** [review]
LGTM! 🚀

**Pull request merged**
